operatorcourier/flatten.py

```diff
@@ -0,0 +1,151 @@
+import logging
+import os
+from yaml import safe_load, MarkedYAMLError
+import semver
+import ntpath
+import operatorcourier.identify as identify
+import operatorcourier.errors as errors
+
+logger = logging.getLogger(__name__)
+
+
+def get_flattened_files_info(source_dir: str) -> [(str, str)]:
+    """
+    :param source_dir: Path of the directory containing different versions
+    of operator bundles (CRD, CSV, package) in separate version directories
+    :return: A list of tuples where in each tuple, the first element is
+    the source file path to be copied to the flattened directory, and the second is
+    the new file name to be used in the file copy. The function returns an empty list if
+    the source_dir is already flat
+    """
+
+    # extract package file and version folders from source_dir
+    _, folder_names, file_names = next(os.walk(source_dir))
+    if len(folder_names) == 0:
+        logger.info('The source directory is already flat. Exiting the program...')
+        return []
+
+    file_paths_to_copy = []  # [ (SRC_FILE_PATH, NEW_FILE_NAME) ]
+
+    crd_dict = {}  # { CRD_NAME => (VERSION, CRD_PATH) }
+    csv_paths = []
+
+    for version_folder_name in folder_names:
+        parse_version_folder(source_dir, version_folder_name, csv_paths, crd_dict)
+
+    # 1. add package in source_dir
+    package_path, package_name = get_package_file(source_dir, file_names)
+    file_paths_to_copy.append((package_path, package_name))
+
+    # 2. add all CRDs with the latest version
+    for crd_name, (_, crd_path) in crd_dict.items():
+        file_paths_to_copy.append((crd_path, ntpath.basename(crd_path)))
+
+    # 3. add all CSVs
+    for csv_file_name, csv_entries in create_csv_dict(csv_paths).items():
+        if len(csv_entries) == 1:
+            file_paths_to_copy.append((csv_entries[0][1], csv_file_name))
+            continue
+        # duplicate CSVs found in different versions
+        for (version, csv_path) in csv_entries:
+            basename, ext = os.path.splitext(csv_file_name)
+            file_paths_to_copy.append((csv_path, f'{basename}_v{version}{ext}'))
+
+    return file_paths_to_copy
+
+
+def parse_version_folder(base_dir: str, version_folder_name: str,
+                         csv_paths: list, crd_dict: (str, str)):
```

Comment:
is `(str, str)` the right way how to describe dict in typing? Shouldn't be this `Dict[str, Tuple[str, str]]` ?

Reply:
also I don't see `mypy` enabled in `tox`

Reply:
tbh, I don't see introducing type annotations as a priority, definitely not in this PR, and I would argue to remove them for now.

If you think type annotations  would improve the code quality, then better document this in the contribution doc, add a check for this in tox and update the whole project to meet the requirement.

Reply:
+1